### PR TITLE
Migrate Counter visualization to React

### PR DESCRIPTION
**Pull request by kravets-levko**
## What type of PR is this? (check all applicable)

- [x] Feature
- [x] Refactor

## Description

- [x] Convert component to React
  - [x] Renderer
  - [x] Editor
- [x] Fix things mentioned in comments (handle edge cases, etc.)
- [x] Cleanup
- [x] Test

## Related Tickets & Documents

getredash/redash#3301 (Migrate Visualizations to React -> Counter)

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)

image
image

### CODE REVIEW

File: client/app/visualizations/counter/utils.js [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+  const result = numeral(value).format(formatString);
+
+  locale.delimiters = savedDelimiters;
+  return result;
+}
+
+// 0 - special case, use first record
+// 1..N - 1-based record number from beginning (wraps if greater than dataset size)
+// -1..-N - 1-based record number from end (wraps if greater than dataset size)
+function getRowNumber(index, rowsCount) {
```
Comment (kravets-levko) [inline client/app/visualizations/counter/utils.js]: @arikfr I may need your opinion about this ^

File: client/app/visualizations/counter/Renderer.jsx [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+import { RendererPropTypes } from '@/visualizations';
+
+import { getCounterData } from './utils';
+
+import './render.less';
+
+function getCounterStyles(scale) {
+  return {
+    OTransform: `scale(${scale})`,
+    MsTransform: `scale(${scale})`,
```
Comment (ranbena) [inline client/app/visualizations/counter/Renderer.jsx]: I believe this should be `msTransform`.

In any case, wdyt of using `modernizr-prefixed` or a simple custom util here instead?

File: client/app/visualizations/counter/Renderer.jsx [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    counterLabel,
+  } = getCounterData(data.rows, options, visualizationName);
+  return (
+    <div
+      className={cx(
+        'counter-visualization-container',
+        { 'trend-positive': showTrend && trendPositive, 'trend-negative': showTrend && !trendPositive },
+      )}
+    >
+      <div className="counter-visualization-content" ref={setContainer}>
```
Comment (ranbena) [inline client/app/visualizations/counter/Renderer.jsx]: Why not use the dedicated `useRef`?

File: client/app/visualizations/counter/Renderer.jsx [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  const [container, setContainer] = useState(null);
+
+  useEffect(() => {
+    if (container) {
+      const unwatch = resizeObserver(container, () => {
+        setScale(getCounterScale(container));
+      });
+      return unwatch;
+    }
+  }, [container, setScale]);
```
Comment (ranbena) [inline client/app/visualizations/counter/Renderer.jsx]: I'm probably not understanding how this works, but are you sure `setScale` is correct in the 2nd argument?

File: client/app/visualizations/counter/Editor.jsx [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+        <GeneralSettings {...props} onOptionsChange={optionsChanged} />
+      </Tabs.TabPane>
+      <Tabs.TabPane key="format" tab={<span data-test="Counter.EditorTabs.Formatting">Format</span>}>
+        <FormatSettings {...props} onOptionsChange={optionsChanged} />
+      </Tabs.TabPane>
+    </Tabs>
+  );
+}
+
+Editor.propTypes = EditorPropTypes;
```
Comment (ranbena) [inline client/app/visualizations/counter/Editor.jsx]: @arikfr wdyt of having a link to the counter video from here?
(Actually, linking to a help doc or blog post with the video in it would be better)

File: client/cypress/integration/visualizations/counter_spec.js [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+    cy.clickThrough(`
+      NewVisualization
+      VisualizationType
+      VisualizationType.COUNTER
+
+      Counter.General.ValueColumn
+      Counter.General.ValueColumn.a
+    `);
+
+    cy.wait(200); // eslint-disable-line cypress/no-unnecessary-waiting
```
Comment (ranbena) [inline client/cypress/integration/visualizations/counter_spec.js]: What's the wait for?

File: client/app/visualizations/counter/Editor.jsx (hunk 5 above)
Comment (arikfr) [reply client/app/visualizations/counter/Editor.jsx]: I agree that having a help trigger here makes sense. :+1: But I guess it should be a global thing for all visualizations, which we enable when we have a relevant page?

File: client/app/visualizations/counter/utils.js [hunk 7]
```diff
@@ -0,0 +1,141 @@
+import { isNumber, isFinite, toString } from 'lodash';
+import numeral from 'numeral';
+
+// TODO: allow user to specify number format string instead of delimiters only
+// It will allow to remove this function (move all that weird formatting logic to a migration
+// that will set number format for all existing counter visualization)
```
Comment (arikfr) [inline client/app/visualizations/counter/utils.js]: Can't we do this already without a migration by just translating the different settings into a format string for existing visualizations and use a string for new ones?

File: client/app/visualizations/counter/utils.js (hunk 1 above)
Comment (arikfr) [reply client/app/visualizations/counter/utils.js]: The description sounds right :)

Comment (kravets-levko) [reply client/app/visualizations/counter/utils.js]: Previous implementation worked in a similar way (I believe), so I first described how it should behave to be clear for users, and then implemented according to that.

File: client/app/visualizations/counter/Renderer.jsx (hunk 2 above)
Comment (kravets-levko) [reply client/app/visualizations/counter/Renderer.jsx]: Yeah, you're right - `-ms-` is the only exception :man_facepalming: Also, probably I'll remove `-o-` rule (Opera 12 is in fact dead)

File: client/app/visualizations/counter/Renderer.jsx (hunk 3 above)
Comment (kravets-levko) [reply client/app/visualizations/counter/Renderer.jsx]: Previously I had some issues passing a regular ref as a dependency to `useEffect` (it just ignored ref changes). This variant works perfectly

File: client/app/visualizations/counter/Renderer.jsx (hunk 4 above)
Comment (kravets-levko) [reply client/app/visualizations/counter/Renderer.jsx]: Yes it's correct - when you use `useEffect`/`useMemo`/`useCallback` you should pass every dependency you'll use in a hook function because the function will capture them and then you'll need to create new function instance that will capture new dependencies values (you may omit some of dependencies or pass extra ones if you know what you do)

File: client/cypress/integration/visualizations/counter_spec.js (hunk 6 above)
Comment (kravets-levko) [reply client/cypress/integration/visualizations/counter_spec.js]: There were some issues similar to other visualizations tests - visualization needed some delay to properly update itself before taking a snapshot.

File: client/app/visualizations/counter/utils.js (hunk 7 above)
Comment (kravets-levko) [reply client/app/visualizations/counter/utils.js]: It's the next thing I'd like to do. But I think it's better to do it as a separate feature (as implementation of getredash/redash#2558), but for a proper implementation in a single pass we need to do something with getredash/redash#3551 - it will allow to get rid of "Formatting String Prefix/Suffix" settings as well.

Comment (arikfr) [reply client/app/visualizations/counter/utils.js]: :+1:

File: client/app/visualizations/counter/Renderer.jsx (hunk 2 above)
Comment (kravets-levko) [reply client/app/visualizations/counter/Renderer.jsx]: `modernizr` seems to be an overkill here :thinking: I'll check if we're using something similar elsewhere - I'll create a custom utility

File: client/app/visualizations/counter/Renderer.jsx (hunk 4 above)
Comment (gabrieldutra) [reply client/app/visualizations/counter/Renderer.jsx]: It still feels to me like you indeed should pass dependencies that you use in the hook function, but only the ones that get changed over time. Also in [docs](https://reactjs.org/docs/hooks-faq.html#is-it-safe-to-omit-functions-from-the-list-of-dependencies) they don't seem to include the setters in the dependency list

Comment (kravets-levko) [reply client/app/visualizations/counter/Renderer.jsx]: @gabrieldutra You're right. It's also mentioned in hooks API reference. Thanks!

File: client/app/visualizations/counter/Editor.jsx (hunk 5 above)
Comment (ranbena) [reply client/app/visualizations/counter/Editor.jsx]: We can embed the video here then https://redash.io/help/user-guide/visualizations/visualization-types#Counter

File: client/cypress/integration/visualizations/counter_spec.js (hunk 6 above)
Comment (ranbena) [reply client/cypress/integration/visualizations/counter_spec.js]: Then perhaps it makes more sense to place it before the snapshot (instead of before the element assertion) 👆

File: client/app/visualizations/counter/Renderer.jsx (hunk 4 above)
Comment (kravets-levko) [reply client/app/visualizations/counter/Renderer.jsx]: Fixed

File: client/cypress/integration/visualizations/counter_spec.js (hunk 6 above)
Comment (kravets-levko) [reply client/cypress/integration/visualizations/counter_spec.js]: @ranbena Done

File: client/app/visualizations/counter/Renderer.jsx (hunk 3 above)
Comment (gabrieldutra) [reply client/app/visualizations/counter/Renderer.jsx]: [This](https://reactjs.org/docs/hooks-faq.html#how-can-i-measure-a-dom-node) seems relevant to cases where the DOM changes and you need to track it in the ref.

**Comment by kravets-levko** [comment]
@arikfr @ranbena @gabrieldutra If you are happy with this PR and have no more comments to add - I'd like to merge it.

File: client/cypress/integration/visualizations/counter_spec.js [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+
+      Counter.General.TargetValueColumn
+      Counter.General.TargetValueColumn.b
+    `);
+
+    cy.getByTestId('VisualizationPreview').find('.counter-visualization-container').should('exist');
+
+    // wait a bit before taking snapshot
+    cy.wait(500); // eslint-disable-line cypress/no-unnecessary-waiting
+    cy.percySnapshot('Visualizations - Counter (format target value)', { widths: [viewportWidth] });
```
Comment (gabrieldutra) [inline client/cypress/integration/visualizations/counter_spec.js]: It seems all the Percy Snapshots have the same name :sweat_smile:, guess it will mess with Percy...

Comment (kravets-levko) [reply client/cypress/integration/visualizations/counter_spec.js]: Uh-oh... Thanks! Probably a copy-paste leftover...

**Comment by kravets-levko** [comment]
> This is probably part of another bug

Not necessarily, it may be introduced in this PR - I'll check

File: client/cypress/integration/visualizations/counter_spec.js (hunk 8 above)
Comment (kravets-levko) [reply client/cypress/integration/visualizations/counter_spec.js]: Fixed :+1:

**Comment by kravets-levko** [comment]
> This is probably part of another bug, but as a note, when you have an empty filter the Counter now disappears.

@gabrieldutra There is a condition - counter will not render anything if there are no rows in result - and this is intended behavior. I think it's kinda correct (probably we can show `0` if `Count rows` option enabled), but we can discuss it and change something in another PR.

**Comment by gabrieldutra** [comment]
It makes sense for it to show 0 when count rows is enabled, but not in this scope indeed.

**Comment by arikfr** [comment]
:+1:
